### ISSUE TITLE
XC_MethodHook callbacks are not working

### ISSUE THREAD

**Issue by MPeti1** [rovo89/Xposed#343]
method hooks don't get called for some reason. i checked in the logs and handleLoadPackage is called, it even reaches the end so findAndHookMethod-s also should be called. in the xposed log there are no related error messages, it just dont work as expected. [here](https://gist.github.com/MPeti1/b322034609d05e0c28789a244a611093) is my code and the logs. other modules (created by others) are working perfectly

**Comment by MPeti1** [comment]
Added this before findAndHookMethod and now the hook is working, but why?

```java
Method onCreate = XposedHelpers.findMethodExact("com.mpeter.asd.MainActivity", lpparam.classLoader, "onCreate", Bundle.class);
        Class declaringClass = onCreate.getDeclaringClass();
        String name = onCreate.getName();
        String str = onCreate.toString();`
```

**Comment by rovo89** [comment]
I think you have hit an edge case with either the caching or the internal hash map. You have used both of these:
```java
XposedHelpers.findAndHookMethod(MainActivity.class, "onCreate", Bundle.class, new XC_MethodHook() ...
XposedHelpers.findAndHookMethod("com.mpeter.asd.MainActivity", lpparam.classLoader, "onCreate", Bundle.class, new XC_MethodHook() ...
```

The difference is that the first will resolve the class with the classloader executing this code (i.e. the one which has loaded your module), whereas the second one uses the classloader of the app (i.e. the GUI). Sounds like that would be the same, but it isn't. Both will look at the same APK/DEX, but each one will return a distinct class object, with seperate static fields, methods, ....

So actually you're hooking two different methods with these two lines. The first one would only be triggered if you called `MainActivity.onCreate()` within your module code, but not when it's called by the Android system during the app lifecycle (again, these are two different methods).

Internally, Xposed has a cache for methods, and I think it might use the class name, method name and signature as key, which doesn't consider the classloader. Not sure how it is for the map which stores hooked methods and the according callbacks. No idea currently whether the comparisons consider the classloader.

**In short: Don't use `xyz.class` for hooking anything in your own app, use `lpparam.classLoader` instead.**
It's ok for Android framework classes though, as they'll always be resolved by the boot classloader.

**Comment by MPeti1** [comment]
So if i hook a method in my own app, i should do it with the string representation of the class name (and the classloader which came with lpparam) ? i think i've tried this before, but I'm not sure, maybe it was once when my module didn't got loaded because i rebooted my phone very soon, and i thought this approach is also not working :D i'll try it out in a minute.
If that's the case, in what cases can i reference the class with the MainActivity.class form? only when i hook classes that are here in every app? what about the libraries, for example the support libraries?

If this was the problem, that means there is a bug where findAndHookMethod can reuse the cached version of a method even if the classloader differs in the new hooked method? (so for example if i do findMethodExact and then findAndHookMethod on seemingly the same method, findAndHookMethod can use the method hooked by findMethodExact?)

**Comment by MPeti1** [comment]
It seems working now with the second way :D 
I forgot to mention i use MIUI 8, and then a modified version of Xposed (by SolarWarez). It's not very up to date (it's version is 87.1) so it can be possible that you already fixed that in the main version
